orca-igor/src/main/groovy/com/netflix/spinnaker/orca/igor/tasks/RetryableIgorTask.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2019 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.orca.igor.tasks;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spinnaker.orca.ExecutionStatus;
+import com.netflix.spinnaker.orca.RetryableTask;
+import com.netflix.spinnaker.orca.TaskResult;
+import com.netflix.spinnaker.orca.pipeline.model.Stage;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import retrofit.RetrofitError;
+
+import javax.annotation.Nonnull;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+@RequiredArgsConstructor
+@Slf4j
+public abstract class RetryableIgorTask implements RetryableTask {
+  public long getBackoffPeriod() {
+    return TimeUnit.SECONDS.toMillis(5);
+  }
+
+  public long getTimeout() {
+    return TimeUnit.MINUTES.toMillis(1);
+  }
+
+  protected int getMaxConsecutiveErrors() {
+    return 5;
+  }
+
+  @Override
+  public @Nonnull TaskResult execute(@Nonnull Stage stage) {
+    CIJobRequest jobRequest = stage.mapTo(CIJobRequest.class);
+    int errors = jobRequest.getConsecutiveErrors();
+    try {
+      TaskResult result = tryExecute(jobRequest);
+      return resetErrorCount(result);
+    } catch (RetrofitError e) {
+      int status = e.getResponse().getStatus();
+      if (jobRequest.getConsecutiveErrors() < getMaxConsecutiveErrors() && isRetryable(status)) {
+        log.warn(String.format("Received HTTP %s response from igor, retrying...", status));
+        return new TaskResult(ExecutionStatus.RUNNING, errorContext(errors + 1));
+      }
+      throw e;
+    }
+  }
+
+  abstract protected @Nonnull TaskResult tryExecute(@Nonnull CIJobRequest ciJobRequest);
+
+  private TaskResult resetErrorCount(TaskResult result) {
+    Map<String, Object> newContext = ImmutableMap.<String, Object>builder()
+      .putAll(result.getContext())
+      .put("consecutiveErrors", 0)
+      .build();
+    return new TaskResult(result.getStatus(), newContext, result.getOutputs());
+  }
+
+  private Map<String, Integer> errorContext(int errors) {
+    return Collections.singletonMap("consecutiveErrors", errors);
+  }
+
+  private boolean isRetryable(int statusCode) {
```

Comment:
I think we might want to retry once or twice on these two error codes. Maybe the error is ephemeral and will fix itself?

Reply:
The two ones defined here (500, 503) are the ones where we do retry; any others (404, etc.) will not be retried.

I got the list from [here](https://github.com/spinnaker/orca/blob/f933bb73f90bde6b04c1e7b2cf2a0fbccdc69bb9/orca-igor/src/main/groovy/com/netflix/spinnaker/orca/igor/tasks/MonitorJenkinsJobTask.groovy#L103) though I dropped 404 as I think that was only in the list to handle cases where the build itself hadn't appeared yet.  (Looks like there were a few fixes to that file specifically to stop continually retrying fetching properties on a 404.)

Reply:
Ha! Great.